src/platform/windows/mod.rs

```diff
@@ -0,0 +1,1482 @@
+{
+    if handle == INVALID_HANDLE_VALUE {
+        return Ok(INVALID_HANDLE_VALUE);
+    }
+
+    unsafe {
+        let mut new_handle: HANDLE = INVALID_HANDLE_VALUE;
+        let ok = kernel32::DuplicateHandle(*CURRENT_PROCESS_HANDLE as HANDLE, handle,
+                                           other_process, &mut new_handle,
+                                           0, winapi::FALSE, flags);
+        if ok == winapi::FALSE {
+            Err(WinError::last("DuplicateHandle"))
+        } else {
+            Ok(new_handle)
+        }
+    }
+}
+
+// duplicate a handle in the current process
+fn dup_handle(handle: HANDLE) -> Result<HANDLE,WinError> {
+    dup_handle_to_process(handle, unsafe { kernel32::GetCurrentProcess() })
+}
+
+// duplicate a handle to the target process
+fn dup_handle_to_process(handle: HANDLE, other_process: HANDLE) -> Result<HANDLE,WinError> {
+    dup_handle_to_process_with_flags(handle, other_process, winapi::DUPLICATE_SAME_ACCESS)
+}
+
+// duplicate a handle to the target process, closing the source handle
+fn move_handle_to_process(handle: HANDLE, other_process: HANDLE) -> Result<HANDLE,WinError> {
+    dup_handle_to_process_with_flags(handle, other_process, winapi::DUPLICATE_CLOSE_SOURCE | winapi::DUPLICATE_SAME_ACCESS)
+}
+
+// get the current process handle as a handle in another process
+#[allow(dead_code)]
+fn current_process_handle_to_process(other_process: HANDLE) -> Result<HANDLE,WinError> {
+    unsafe {
+        let mut new_handle: HANDLE = INVALID_HANDLE_VALUE;
+        let ok = kernel32::DuplicateHandle(kernel32::GetCurrentProcess(), kernel32::GetCurrentProcess(),
+                                           other_process, &mut new_handle,
+                                           winapi::PROCESS_DUP_HANDLE, winapi::FALSE, 0);
+        if ok == winapi::FALSE {
+            Err(WinError::last("DuplicateHandle"))
+        } else {
+            Ok(new_handle)
+        }
+    }
+}
+
+fn create_overlapped() -> Box<winapi::OVERLAPPED> {
+    Box::new(winapi::OVERLAPPED {
+        Internal: 0,
+        InternalHigh: 0,
+        Offset: 0,
+        OffsetHigh: 0,
+        hEvent: 0 as HANDLE
+    })
+}
+
+fn reset_overlapped(ov: &mut winapi::OVERLAPPED) {
+    ov.Internal = 0;
+    ov.InternalHigh = 0;
+    ov.Offset = 0;
+    ov.OffsetHigh = 0;
+    ov.hEvent = 0 as HANDLE;
+}
+
+
+#[derive(Debug, PartialEq)]
+struct WinHandle {
+    h: HANDLE
+}
+
+unsafe impl Send for WinHandle { }
+unsafe impl Sync for WinHandle { }
+
+impl Drop for WinHandle {
+    fn drop(&mut self) {
+        self.close();
+    }
+}
+
+impl Default for WinHandle {
+    fn default() -> WinHandle {
+        WinHandle { h: INVALID_HANDLE_VALUE }
+    }
+}
+
+impl Deref for WinHandle {
+    type Target = HANDLE;
+
+    #[inline]
+    fn deref(&self) -> &HANDLE {
+        &self.h
+    }
+}
+
+impl WinHandle {
+    fn new(h: HANDLE) -> WinHandle {
+        WinHandle { h: h }
+    }
```

Comment:
Exactly.

(Side note on terminology: buggy code that could actually trigger undefined behaviour is called "unsound" in Rust; as opposed to "unsafe", which just indicates code for which the compiler can't guarantee soundness, and thus the burden is on the programmer...)

I actually don't know what best practices are regarding code that shouldn't be able to cause unsoundness, but can trigger other errors that would normally be prevented by Rust's type system, in the absence of unsafe operations... On one hand, since its correctness is affected by the presence of unsafe blocks, it is tempting to flag it as unsafe as well; on the other hand, this might dilute attention from the code that might actually be unsound...

Either way, since this doesn't actually affect soundness, I no longer see it as a major concern :-)